src/main/java/com/google/api/codegen/discovery/ViewModelProvider.java

```diff
@@ -0,0 +1,191 @@
+      ApiaryConfig apiaryConfig,
+      CommonSnippetSetRunner snippetSetRunner,
+      SampleMethodToViewTransformer methodToViewTransformer,
+      JsonNode sampleConfigOverrides,
+      TypeNameGenerator typeNameGenerator,
+      String outputRoot) {
+    this.apiaryConfig = apiaryConfig;
+    this.snippetSetRunner = snippetSetRunner;
+    this.methodToViewTransformer = methodToViewTransformer;
+    this.sampleConfigOverrides = sampleConfigOverrides;
+    this.typeNameGenerator = typeNameGenerator;
+    this.outputRoot = outputRoot;
+  }
+
+  @Override
+  public Map<String, Doc> generate(Method method) {
+    // Before the transformer step, we generate the SampleConfig and apply overrides if available.
+    SampleConfig sampleConfig =
+        new ApiaryConfigToSampleConfigConverter(
+                Collections.singletonList(method), apiaryConfig, typeNameGenerator)
+            .convert();
+    sampleConfig = override(sampleConfig, sampleConfigOverrides);
+    ViewModel surfaceDoc = methodToViewTransformer.transform(method, sampleConfig);
+    Doc doc = snippetSetRunner.generate(surfaceDoc);
+    Map<String, Doc> docs = new TreeMap<>();
+    if (doc == null) {
+      return docs;
+    }
+    docs.put(outputRoot + "/" + surfaceDoc.outputPath(), doc);
+    return docs;
+  }
+
+  /**
+   * Applies sampleConfigOverrides to sampleConfig.
+   *
+   * If sampleConfigOverrides is null, sampleConfig is returned as is.
+   */
+  private static SampleConfig override(SampleConfig sampleConfig, JsonNode sampleConfigOverrides) {
+    // We use JSON merging to facilitate this override mechanism:
+    // 1. Convert the SampleConfig into a JSON tree.
+    // 2. Convert the overrides file into a JSON tree with arbitrary schema.
+    // 3. Overwrite object fields of the SampleConfig tree where field names match.
+    // 4. Convert the modified SampleConfig tree back into a SampleConfig.
+    if (sampleConfigOverrides != null) {
+      ObjectMapper mapper = new ObjectMapper();
+      JsonNode tree = mapper.valueToTree(sampleConfig);
+      merge((ObjectNode) tree, (ObjectNode) sampleConfigOverrides);
+      try {
+        return mapper.treeToValue(tree, SampleConfig.class);
+      } catch (Exception e) {
+        throw new RuntimeException("failed to parse config to node: " + e.getMessage());
+      }
+    }
+    return sampleConfig;
+  }
+
+  /**
+   * Overwrites the fields of tree that intersect with those of overrideTree.
+   *
+   * Values of tree are modified, and values of overrideTree are not modified.
+   * The merge process loops through the fields of tree and replaces non-object
+   * values with the corresponding value from overrideTree if present. Object
+   * values are traversed recursively to replace sub-properties.
+   *
+   * @param tree the original JsonNode to modify.
+   * @param overrideTree the JsonNode with values to overwrite tree with.
+   */
+  private static void merge(ObjectNode tree, ObjectNode overrideTree) {
```

Comment:
sg. So, 
(TL;DR)^2: each component that we expect could possibly be overridden only appears once in the overrides file/`SampleConfig`. The actual identifiers that we construct in the samples are built from these components. Correct?

It would be good to capture the explanation you provided somewhere obvious. Maybe as part of the class comment of the Discovery `SampleConfig`?